swap/swap.go

```diff
@@ -1,98 +1,636 @@
+		disconnectThreshold: DefaultDisconnectThreshold,
+		oracle:              NewPriceOracle(),
 	}
-	return
 }
 
-//Swap implements the protocols.Balance interface
-//Add is the (sole) accounting function
+const (
+	balancePrefix        = "balance_"
+	sentChequePrefix     = "sent_cheque_"
+	receivedChequePrefix = "received_cheque_"
+)
+
+// returns the store key for retrieving a peer's balance
+func balanceKey(peer enode.ID) string {
+	return balancePrefix + peer.String()
+}
+
+// returns the store key for retrieving a peer's last sent cheque
+func sentChequeKey(peer enode.ID) string {
+	return sentChequePrefix + peer.String()
+}
+
+// returns the store key for retrieving a peer's last received cheque
+func receivedChequeKey(peer enode.ID) string {
+	return receivedChequePrefix + peer.String()
+}
+
+func keyToID(key string, prefix string) enode.ID {
+	return enode.HexID(key[len(prefix):])
+}
+
+// createOwner assings keys and addresses
+func createOwner(prvkey *ecdsa.PrivateKey, contract common.Address) *Owner {
+	pubkey := &prvkey.PublicKey
+	return &Owner{
+		privateKey: prvkey,
+		publicKey:  pubkey,
+		Contract:   contract,
+		address:    crypto.PubkeyToAddress(*pubkey),
+	}
+}
+
+// DeploySuccess is for convenience log output
+func (s *Swap) DeploySuccess() string {
+	return fmt.Sprintf("contract: %s, owner: %s, deposit: %v, signer: %x", s.owner.Contract.Hex(), s.owner.address.Hex(), s.params.InitialDepositAmount, s.owner.publicKey)
+}
+
+// Add is the (sole) accounting function
+// Swap implements the protocols.Balance interface
 func (s *Swap) Add(amount int64, peer *protocols.Peer) (err error) {
-	s.lock.Lock()
-	defer s.lock.Unlock()
+	s.accountingLock.Lock()
+	defer s.accountingLock.Unlock()
 
-	//load existing balances from the state store
-	err = s.loadState(peer)
+	err = s.loadBalance(peer.ID())
 	if err != nil && err != state.ErrNotFound {
-		return
+		return fmt.Errorf("error while loading balance for peer %s", peer.ID().String())
+	}
+
+	// Check if balance with peer is over the disconnect threshold
+	balance, exists := s.getBalance(peer.ID())
+	if !exists {
+		return fmt.Errorf("peer %v does not exist", peer.ID())
+	}
+	if balance >= s.disconnectThreshold {
+		return fmt.Errorf("balance for peer %s is over the disconnect threshold %d, disconnecting", peer.ID().String(), s.disconnectThreshold)
+	}
+
+	var newBalance int64
+	newBalance, err = s.updateBalance(peer.ID(), amount)
+	if err != nil {
+		return err
+	}
+
+	// Check if balance with peer crosses the payment threshold
+	// It is the peer with a negative balance who sends a cheque, thus we check
+	// that the balance is *below* the threshold
+	if newBalance <= -s.paymentThreshold {
+		log.Warn("balance for peer went over the payment threshold, sending cheque", "peer", peer.ID().String(), "payment threshold", s.paymentThreshold)
+		return s.sendCheque(peer.ID())
 	}
+
+	return nil
+}
+
+func (s *Swap) getBalance(id enode.ID) (int64, bool) {
+	s.balancesLock.RLock()
```

Comment:
the `Balance` function (`swap.go:182`) uses `getBalance` outside of `Add`, so in this case we wouldn't be under the lock applied by `accountingLock`.

in the case of cheques and peers, i think it's reasonable to expect eventual access to these maps under a different context. but we _could_ take the approach of deleting these until we actually need them.

regarding the loading: not sure what you mean here. should we be accessing the store even in the context of updating balances and creating cheques? wouldn't this slow things down unnecessarily?

(if possible, i would like to request @holisticode's and @acud's POV on these matters)

Reply:
Again @zelig, we may have misunderstood and misapplied @acud 's review suggestions, but what we understood was that accesses to the maps should not be done directly and be under a lock, and he insisted that his PR approval would be linked to redesigning the locks. 

We were always aware that `accountingLock` is securing the whole `Add` run and we were safe.

Reply:
@mortelli @holisticode if you want to use getters and setters - i think they should under a lock just the way they are here.
that being said i do feel you could get rid of the getters and setters altogether and just handle the maps directly under the accounting lock in the methods that actually do the business logic. it would make the behavior of the code easier to read and understand. IMO when using getters and setters, having locks on just some of them is kind of inviting trouble and future deadlocks and quirks are possibly introduced.

Reply:
relevant issue: #1663.